File: yt/core/ypath/tokenizer.cpp
#include "tokenizer.h"

#include <core/misc/error.h>
#include <core/misc/string.h>

#include <core/yson/token.h>

namespace NYT {
namespace NYPath {

////////////////////////////////////////////////////////////////////////////////

TTokenizer::TTokenizer(const TYPath& path)
{
    Path_ = path;
    Input_ = Path_;
    Token_ = TStringBuf();
    Type_  = ETokenType::StartOfStream;
    LiteralValue_.reserve(Path_.length());
}

ETokenType TTokenizer::Advance()
{
    // Replace Input_ with suffix.
    Input_ = TStringBuf(Input_.begin() + Token_.length(), Input_.end());
    LiteralValue_.clear();

    // Check for EndOfStream.
    const char* current = Input_.begin();
    if (current == Input_.end()) {
        Type_ = ETokenType::EndOfStream;
        return Type_;
    }

    Type_ = ETokenType::Literal;
    bool proceed = true;
    while (proceed && current != Input_.end()) {
        auto token = NYson::CharToTokenType(*current);
        if (token == NYson::ETokenType::LeftBracket ||
            token == NYson::ETokenType::LeftBrace)
        {
            if (current == Input_.begin()) {
                Type_ = ETokenType::Range;
                current = Input_.end();
            }
            proceed = false;
            continue;
        }

        switch (*current) {
            case '/':
            case '@':
            case '&':
            case '*':
                if (current == Input_.begin()) {
                    Token_ = TStringBuf(current, current + 1);
                    switch (*current) {
                        case '/': Type_ = ETokenType::Slash;     break;
                        case '@': Type_ = ETokenType::At;        break;
                        case '&': Type_ = ETokenType::Ampersand; break;
                        case '*': Type_ = ETokenType::Asterisk;  break;
                        default:  YUNREACHABLE();
                    }
                    return Type_;
                }
                proceed = false;
                break;

            case '\\':
                current = AdvanceEscaped(current);
                break;

            default:
                LiteralValue_.append(*current);
                ++current;
                break;
        }
    }

    Token_ = TStringBuf(Input_.begin(), current);
    return Type_;
}

void TTokenizer::ThrowMalformedEscapeSequence(const TStringBuf& context)
{
    THROW_ERROR_EXCEPTION("Malformed escape sequence %Qv",
        context);
}

const char* TTokenizer::AdvanceEscaped(const char* current)
{
    YASSERT(*current == '\\');
    ++current;

    if (current == Input_.end()) {
        THROW_ERROR_EXCEPTION("Premature end-of-string while parsing escape sequence");
    }

    switch (*current) {
        case '\\':
        case '/':
        case '@':
        case '&':
        case '*':
        case '[':
        case '{':
            LiteralValue_.append(*current);
            ++current;
            break;

        case 'x': {
            if (current + 2 >= Input_.end()) {
                ThrowMalformedEscapeSequence(TStringBuf(current - 1, Input_.end()));
            }
            TStringBuf context(current - 1, current + 3);
            int hi = ParseHexDigit(current[1], context);
            int lo = ParseHexDigit(current[2], context);
            LiteralValue_.append((hi << 4) + lo);
            current = context.end();
            break;
        }

        default:
            ThrowMalformedEscapeSequence(TStringBuf(current - 1, current + 1));
    }

    return current;
}

int TTokenizer::ParseHexDigit(char ch, const TStringBuf& context)
{
    if (ch >= '0' && ch <= '9') {
        return ch - '0';
    }

    if (ch >= 'a' && ch <= 'f') {
        return ch - 'a' + 10;
    }

    if (ch >= 'A' && ch <= 'F') {
        return ch - 'A' + 10;
    }

    ThrowMalformedEscapeSequence(context);
    YUNREACHABLE();
}

void TTokenizer::Expect(ETokenType expectedType)
{
    if (expectedType != Type_) {
        if (Type_ == ETokenType::EndOfStream) {
            THROW_ERROR_EXCEPTION("Premature end-of-stream while expecting %Qlv",
                expectedType);
        } else {
            THROW_ERROR_EXCEPTION("Expected %Qlv but found %Qlv token %Qv",
                expectedType,
                Type_,
                Token_);
        }
    }
}

void TTokenizer::ThrowUnexpected()
{
    if (Type_ == ETokenType::EndOfStream) {
        THROW_ERROR_EXCEPTION("Unexpected end-of-stream");
    } else {
        THROW_ERROR_EXCEPTION("Unexpected %Qlv token %Qv",
            Type_,
            Token_);
    }
}

ETokenType TTokenizer::GetType() const
{
    return Type_;
}

const TStringBuf& TTokenizer::GetToken() const
{
    return Token_;
}

TYPath TTokenizer::GetPrefix() const
{
    return TYPath(Path_.begin(), Input_.begin());
}

TYPath TTokenizer::GetSuffix() const
{
    return TYPath(Input_.begin() + Token_.length(), Input_.end());
}

TYPath TTokenizer::GetInput() const
{
    return TYPath(Input_);
}

const Stroka& TTokenizer::GetLiteralValue() const
{
    return LiteralValue_;
}

////////////////////////////////////////////////////////////////////////////////

} // namespace NYPath
} // namespace NYT

File: yt/python/yt/ypath/tokenizer.py
from .common import YPathError

from yt.yson.yson_token import (
    TOKEN_RANGE,
    TOKEN_LITERAL,
    TOKEN_START_OF_STREAM,
    TOKEN_END_OF_STREAM,
    TOKEN_LEFT_BRACE,
    TOKEN_LEFT_BRACKET,
    char_to_token_type)


class YPathTokenizer(object):
    def __init__(self, path):
        self._path = path
        self._type = TOKEN_START_OF_STREAM
        self._token = b""

        self._prefix_len = 0
        self._offset = 0

    def advance(self):
        self._offset += len(self._token)

        current = self._offset
        if current == len(self._path):
            self._type = TOKEN_END_OF_STREAM
            return self._type

        self._type = TOKEN_LITERAL
        while current != len(self._path):
            token = char_to_token_type(self._path[current])
            if token == TOKEN_LEFT_BRACE or token == TOKEN_LEFT_BRACKET or self._path[current] in b"/@&*":
                break

            elif self._path[current] in b"\\":
                current = self.advance_escaped(current)
            else:
                current += 1

        if current == self._offset:
            token = char_to_token_type(self._path[current])
            if token == TOKEN_LEFT_BRACE or token == TOKEN_LEFT_BRACKET:
                self._type = TOKEN_RANGE
                current = len(self._path)
            elif self._path[current] in b"/@&*":
                self._token = self._path[current:current + 1]
                self._type = char_to_token_type(self._path[current])
                return self._type

        self._token = self._path[self._offset: current]
        return self._type

    def throw_malformed_escape_sequence(self, context):
        raise YPathError("Malformed escape sequence {0} in YPath".format(context))

    def advance_escaped(self, current):
        assert self._path[current] in b"\\"
        current += 1
        if current == len(self._path):
            raise YPathError("Unexpected end-of-string in YPath while parsing escape sequence")

        if self._path[current] in b"\\/@&*[{":
            current += 1
        elif self._path[current] in b"x":
            if current + 2 >= len(self._path):
                self.throw_malformed_escape_sequence(self._path[current - 1:])
            context = self._path[current - 1: current + 2]
            try:
                int(self._path[current + 1: current + 3], base=16)
                current += 3
            except ValueError:
                self.throw_malformed_escape_sequence(context)
        else:
            self.throw_malformed_escape_sequence(self._path[current - 1:current + 1])
        return current

    def skip(self, type):
        if self._type == type:
            self.advance()
            return True
        return False

    def get_type(self):
        return self._type

    def get_token(self):
        return self._token

    def get_prefix(self):
        return self._path[:self._offset]
